### PR TITLE
[dicp][ascend]infer op resinfo_231122_1

### PR DESCRIPTION
**Pull request by KevinfromTJ**
1. finish some op's result inference : **softmax+log_softmax+add+amax(keepdim=True)**
2. pass **static** test

### CODE REVIEW

File: dipu/third_party/DIOPI
Comment (zhaochaoxing) [inline dipu/third_party/DIOPI:1]: 非必要不更新DIOPI

File: dicp/dicp/dynamo_bridge/graph.py [hunk 1]
```diff
@@ -35,17 +36,41 @@ def __init__(
     def transform(self):
         self.gm = self.backend_opset_transform(self.gm)
 
+    # def infer_shape_dtype(self):
```
Comment (zhaochaoxing) [inline dicp/dicp/dynamo_bridge/graph.py:39]: 把不需要的注释删掉

File: dicp/dicp/dynamo_bridge/graph.py [hunk 2]
```diff
@@ -56,7 +81,21 @@ def make_tensor_meta(x) -> Optional[TensorMetadata]:
...
                     attr_size, attr_dtye = attr_itr.shape, attr_itr.dtype
                 with FakeTensorMode():
-                    n.meta['val'] = torch.empty(attr_size, dtype=attr_dtye)
+                    fake_value = torch.empty(attr_size, dtype=attr_dtye)
+            else:
+                continue
+            if 'val' in n.meta and test_infer:
+                # import pdb
+                # pdb.set_trace()
+                print("\noperator:\n",n.target)
```
Comment (zhaochaoxing) [inline dicp/dicp/dynamo_bridge/graph.py:90]: 把不需要的print删掉

**Review by zhaochaoxing — APPROVED** [review]

File: dicp/dicp/vendor/AscendGraph/infer_res_utils.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        t2_idx = complex_list.index(type2) if type2 in complex_list else -1
+        return complex_list[max(t1_idx, t2_idx)]
+
+    elif type1 == torch.double or type2 == torch.double:
+        return torch.double
+    elif type1 in float_list or type2 in float_list:
+        t1_idx = float_list.index(type1) if type1 in float_list else -1
+        t2_idx = float_list.index(type2) if type2 in float_list else -1
+        return float_list[max(t1_idx, t2_idx)]
+    elif type1 == torch.bool or type2 == torch.bool:
```
Comment (jinminxi104) [inline dicp/dicp/vendor/AscendGraph/infer_res_utils.py:72]: 我测试了下 bool + int = int
这里的bool和int应该反一下？

**Review by jinminxi104 — CHANGES_REQUESTED** [review]

**Comment by jinminxi104** [comment]
整个推shape的逻辑参考了什么吗? pytorch的那个infer_shape_lib参考了吗？
有些逻辑我担心还是会遗漏

**Comment by zhaochaoxing** [comment]
> 整个推shape的逻辑参考了什么吗? pytorch的那个infer_shape_lib参考了吗？ 有些逻辑我担心还是会遗漏

**Comment by KevinfromTJ** [comment]
> 整个推shape的逻辑参考了什么吗? pytorch的那个infer_shape_lib参考了吗？ 有些逻辑我担心还是会遗漏

大部分有参考的 [华为Ascend PyTorch adapter的推导 （c++）](https://gitee.com/ascend/op-plugin/blob/master/op_plugin/utils/KernelNpuOutputSize.cpp)。

- 复杂的算子各有各的推导，现在只实现了简单的几个和比较通用的。

- 一些很特殊情况没有加入（比如华为源码中对空张量貌似有特殊处理）

File: dicp/dicp/vendor/AscendGraph/infer_res_utils.py (hunk 3 above)
Comment (KevinfromTJ) [reply dicp/dicp/vendor/AscendGraph/infer_res_utils.py]: 已修改

**Comment by zhaochaoxing** [comment]
> 整个推shape的逻辑参考了什么吗? pytorch的那个infer_shape_lib参考了吗？ 有些逻辑我担心还是会遗漏

conversion之后图是由华为算子构成，推华为算子的shape主要参考华为pytorch adaptor比较合适。另外补了check逻辑，conversion相当于一个aten算子的功能拆成多个华为算子组合实现，那么每组华为中算子最后一个算子的输出shape就和与之对应的aten算子的输出shape相同。

File: dipu/scripts/ci/ci_one_iter.sh [hunk 4]
```diff
@@ -42,7 +42,7 @@ function clone_needed_repo() {
     SMART_VERSION=dev_for_mmcv2.0
     MMYOLO=dipu_v0.5.0_one_iter_tool
     DIENGINE=dipu_v0.4.8_one_iter_tool
-    TRANSFORMERS=dipu_v4.28.1_one_iter_tool
+    # TRANSFORMERS=dipu_v4.28.1_one_iter_tool
```
Comment (jinminxi104) [inline dipu/scripts/ci/ci_one_iter.sh:45]: 不能改dipu里面的内容

**Review by jinminxi104 — CHANGES_REQUESTED** [review]

File: dipu/scripts/ci/ci_one_iter.sh (hunk 4 above)
Comment (KevinfromTJ) [reply dipu/scripts/ci/ci_one_iter.sh:45]: 这个没注意，rebase到main上的时候自动加上了

在 https://github.com/DeepLink-org/deeplink.framework/pull/451 看已经merge了，而且最新的dipu好像就是这样的，不知道这里为什么还显示有差异

File: dicp/dicp/vendor/AscendGraph/infer_res_utils.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    elif type1 in int_list or type2 in int_list:
+        t1_idx = int_list.index(type1) if type1 in int_list else -1
+        t2_idx = int_list.index(type2) if type2 in int_list else -1
+        return int_list[max(t1_idx, t2_idx)]
+
+    assert False, str(type1) + " " + str(type2) + " can't cast these two types!"
+    return None
+
+
+def analyze_memory_format(tensor: torch.Tensor, operation: str) -> torch.memory_format:
```
[the remark this answers is not in the document]
Comment (KevinfromTJ) [reply dicp/dicp/vendor/AscendGraph/infer_res_utils.py:84]: 对

File: dipu/scripts/ci/ci_one_iter.sh (hunk 4 above)
Comment (zhaochaoxing) [reply dipu/scripts/ci/ci_one_iter.sh:45]: 估计是rebase的方法不对

Comment (jinminxi104) [reply dipu/scripts/ci/ci_one_iter.sh:45]: 简单merge下主分支也可以的，现在最后都是squash的方式了

**Review by jinminxi104 — APPROVED** [review]

**Review by zhaochaoxing — APPROVED** [review]
